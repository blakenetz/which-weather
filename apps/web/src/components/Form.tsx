import { Search } from "@mui/icons-material";
import {
  Box,
  BoxProps,
  Menu,
  menuClasses,
  MenuItem,
  OutlinedInput,
  styled,
  Typography,
} from "@mui/material";
import { debounce } from "lodash";
import React from "react";
import { WeatherLocation } from "@server/types";
import { ErrorContext } from "@web/context/ErrorContext";
import { locationData } from "@web/test/data";

const locationURL = new URL("location", import.meta.env.VITE_SERVER).toString();
const forecastURL = new URL("forecast", import.meta.env.VITE_SERVER).toString();

const StyledMenu = styled(Menu, {
  shouldForwardProp: (propName) => propName !== "inputWidth",
})<{ inputWidth?: number }>(({ inputWidth }) => ({
  [`.${menuClasses.paper}`]: {
    backgroundColor: "purple",
  },
}));

export default function Form(props: BoxProps<"form">) {
  const errorCtx = React.useContext(ErrorContext);

  const inputRef = React.useRef<HTMLInputElement>(null);
  const formRef = React.useRef<HTMLFormElement>(null);

  const [options, setOptions] = React.useState<WeatherLocation[]>([]);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const debouncedOnChange = React.useCallback(debounce(fetchData, 500), []);

  function fetchData(e: React.ChangeEvent<HTMLInputElement>) {
    const { name, value } = e.target;

    if (!value) return;

    return setOptions(locationData);

    // fetch(locationURL, {
    //   method: "POST",
    //   body: JSON.stringify({ [name]: value }),
    //   headers: { "Content-Type": "application/json" },
    // })
    //   .then(async (res) => {
    //     if (res.ok) setOptions((await res.json()) ?? []);
    //     else errorCtx.setError("autocomplete");
    //   })
    //   .catch(() => errorCtx.setError("location"));
  }

  function handleSubmit(e: React.FormEvent<HTMLFormElement>) {
    e.preventDefault();
    debouncedOnChange.cancel();

    fetch(forecastURL, {
      method: "POST",
      body: new FormData(e.currentTarget),
    })
      .then(async (res) => {
        if (res.status === 200) setOptions((await res.json()) ?? []);
        else errorCtx.setError("forecast");
      })
      .catch(() => errorCtx.setError("unknown"));
  }

  return (
    <Box
      {...props}
      onSubmit={handleSubmit}
      component="form"
      ref={formRef}
      sx={({ breakpoints }) => ({
        marginTop: "15dvh",
        width: "100%",
        maxWidth: breakpoints.values["sm"],
      })}
    >
      <OutlinedInput
        ref={inputRef}
        id="search"
        name="q"
        placeholder="Search by Zip Code or City"
        aria-label="Search by zip code or city"
        autoFocus
        startAdornment={<Search />}
        onChange={debouncedOnChange}
        fullWidth
      />
      <Menu
        id="options-menu"
        anchorEl={inputRef.current!}
        open={Boolean(options.length)}
        onClose={() => setOptions([])}
        MenuListProps={{ "aria-labelledby": "search" }}
        sx={({ breakpoints }) => ({
          [`.${menuClasses.paper}`]: { width: breakpoints.values["sm"] },
        })}
      >
        {options.map((o) => (
          <MenuItem
            key={o.key}
            onClick={() => formRef.current?.requestSubmit()}
            sx={{ gap: 1 }}
          >
            <Typography variant="button" color="primary">
              {o.city}
            </Typography>
            <Typography
              variant="body2"
              color="textSecondary"
            >{`${o.state}, ${o.country}`}</Typography>
          </MenuItem>
        ))}
      </Menu>
    </Box>
  );
}
