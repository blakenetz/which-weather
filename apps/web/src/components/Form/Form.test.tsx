import { expect, test } from "vitest";
import { render } from "@testing-library/react";
import Form from "./Form";

test("Form", async () => {
  const { baseElement } = render(<Form />);

  expect(baseElement).toMatchSnapshot();
});
