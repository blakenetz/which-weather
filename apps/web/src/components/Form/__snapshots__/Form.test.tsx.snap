// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Form 1`] = `
<body>
  <div>
    <form
      class="MuiBox-root css-qzc17b"
    >
      <div
        aria-label="Search by zip code or city"
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-adornedStart css-1rs2f26-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="SearchIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
          />
        </svg>
        <input
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart css-hyt7bp-MuiInputBase-input-MuiOutlinedInput-input"
          id="search"
          name="q"
          placeholder="Search by Zip Code or City"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-12rqw02-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-w4cd9x"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </form>
  </div>
</body>
`;
